cpp/src/rolling/rolling.cu

```diff
@@ -623,5 +625,216 @@ std::unique_ptr<column> rolling_window(column_view const& input,
+                    "Unexpected data-type for timestamp-based rolling window operation!");
+        return 24L*60*60*1000*1000*1000;
+    }
+  }
+
+  template <typename TimestampImpl_t>
+  std::unique_ptr<column> grouped_time_range_rolling_window_impl( column_view const& input,
+                                                      column_view const& timestamp_column,
+                                                      rmm::device_vector<cudf::size_type> const& group_offsets,
+                                                      rmm::device_vector<cudf::size_type> const& group_labels,
+                                                      size_type preceding_window_in_days, // TODO: Consider taking offset-type as type_id. Assumes days for now.
+                                                      size_type following_window_in_days,
+                                                      size_type min_periods,
+                                                      std::unique_ptr<aggregation> const& aggr,
+                                                      rmm::mr::device_memory_resource* mr) {
+
+    TimestampImpl_t mult_factor {static_cast<TimestampImpl_t>(multiplication_factor(timestamp_column.type()))};
+  
+    auto preceding_calculator = 
+      [
+        d_group_offsets = group_offsets.data().get(),
+        d_group_labels  = group_labels.data().get(),
+        d_timestamps    = timestamp_column.data<TimestampImpl_t>(),
+        preceding_window_in_days,
+        mult_factor
+      ] __device__ (size_type idx) {
+        auto group_label = d_group_labels[idx];
+        auto group_start = d_group_offsets[group_label];
+        auto lower_bound = d_timestamps[idx] - preceding_window_in_days*mult_factor;
+
+        auto preceding_i{idx}; // Start at current row.
+        // If preceding value is part of current group, and also exceeds lower_bound, extend window backwards.
+        while ((preceding_i-1) >= group_start && d_timestamps[preceding_i-1] >= lower_bound) {
+          --preceding_i;
+        }
+
+        return idx - preceding_i + 1; // Preceding must account for current row.
+      };
+  
+    auto following_calculator = 
+      [
+        d_group_offsets = group_offsets.data().get(),
+        d_group_labels  = group_labels.data().get(),
+        d_timestamps    = timestamp_column.data<TimestampImpl_t>(),
+        following_window_in_days,
+        mult_factor
+      ] __device__ (size_type idx) {
+        auto group_label = d_group_labels[idx];
+        auto group_end = d_group_offsets[group_label+1]; // Cannot fall off the end, since offsets is capped with `input.size()`.
+        auto upper_bound = d_timestamps[idx] + following_window_in_days*mult_factor;
+
+        auto following_i{idx}; // Start at current row.
+        // If following value is part of current group, and does not exceed upper_bound, extend window forwards.
+        while ((following_i+1) < group_end && d_timestamps[following_i+1] <= upper_bound) {
```

Comment:
See binary search comment above. Here it's `thrust::upper_bound` with forward iterators.

Reply:
Thank you, I've changed this and tested that it works well at scale. I'll update this branch, after I've sorted out the documentation.